crates/bevy_pbr/src/render/pbr_functions.wgsl

```diff
@@ -0,0 +1,196 @@
+    world_tangent: vec4<f32>,
+#endif
+#endif
+    uv: vec2<f32>,
+    is_front: bool,
+) -> vec3<f32> {
+    var N: vec3<f32> = normalize(world_normal);
+
+#ifdef VERTEX_TANGENTS
+#ifdef STANDARDMATERIAL_NORMAL_MAP
+    // NOTE: The mikktspace method of normal mapping explicitly requires that these NOT be
+    // normalized nor any Gram-Schmidt applied to ensure the vertex normal is orthogonal to the
+    // vertex tangent! Do not change this code unless you really know what you are doing.
+    // http://www.mikktspace.com/
+    var T: vec3<f32> = world_tangent.xyz;
+    var B: vec3<f32> = world_tangent.w * cross(N, T);
+#endif
+#endif
+
+    if ((material.flags & STANDARD_MATERIAL_FLAGS_DOUBLE_SIDED_BIT) != 0u) {
+        if (!is_front) {
+            N = -N;
+#ifdef VERTEX_TANGENTS
+#ifdef STANDARDMATERIAL_NORMAL_MAP
+            T = -T;
+            B = -B;
+#endif
+#endif
+        }
+    }
+
+#ifdef VERTEX_TANGENTS
+#ifdef STANDARDMATERIAL_NORMAL_MAP
+    // Nt is the tangent-space normal.
+    var Nt: vec3<f32>;
+    if ((material.flags & STANDARD_MATERIAL_FLAGS_TWO_COMPONENT_NORMAL_MAP) != 0u) {
+        // Only use the xy components and derive z for 2-component normal maps.
+        Nt = vec3<f32>(textureSample(normal_map_texture, normal_map_sampler, uv).rg * 2.0 - 1.0, 0.0);
+        Nt.z = sqrt(1.0 - Nt.x * Nt.x - Nt.y * Nt.y);
+    } else {
+        Nt = textureSample(normal_map_texture, normal_map_sampler, uv).rgb * 2.0 - 1.0;
+    }
+    // Normal maps authored for DirectX require flipping the y component
+    if ((material.flags & STANDARD_MATERIAL_FLAGS_FLIP_NORMAL_MAP_Y) != 0u) {
+        Nt.y = -Nt.y;
+    }
+    // NOTE: The mikktspace method of normal mapping applies maps the tangent-space normal from
+    // the normal map texture in this way to be an EXACT inverse of how the normal map baker
+    // calculates the normal maps so there is no error introduced. Do not change this code
+    // unless you really know what you are doing.
+    // http://www.mikktspace.com/
+    N = normalize(Nt.x * T + Nt.y * B + Nt.z * N);
+#endif
+#endif
+
+    return N;
+}
+
+// NOTE: Correctly calculates the view vector depending on whether
+// the projection is orthographic or perspective.
+fn calculate_view(
+    world_position: vec4<f32>,
+    is_orthographic: bool,
+) -> vec3<f32> {
+    var V: vec3<f32>;
+    if (is_orthographic) {
+        // Orthographic view vector
+        V = normalize(vec3<f32>(view.view_proj[0].z, view.view_proj[1].z, view.view_proj[2].z));
+    } else {
+        // Only valid for a perpective projection
+        V = normalize(view.world_position.xyz - world_position.xyz);
+    }
+    return V;
+}
+
+struct PbrInput {
+    material: StandardMaterial;
+    occlusion: f32;
+    frag_coord: vec4<f32>;
+    world_position: vec4<f32>;
+    world_normal: vec3<f32>;
+    N: vec3<f32>;
+    V: vec3<f32>;
+    is_orthographic: bool;
+};
+
+fn pbr(
+    in: PbrInput,
+) -> [[location(0)]] vec4<f32> {
```

Comment:
Why does this set `[[location(0)]]` on the return? Is that actually used?

Reply:
Good catch! Copy pasta. Fixed.